src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.ChangeNamespaceCodeAction.cs

```diff
@@ -0,0 +1,549 @@
+                //          {
+                //               C2 _c2;    // C2 is define in namespace "Foo.Bar" in another document.
+                //          }
+                //      }
+                //
+                // 3. Change namespace declaration to target namespace.
+                // 4. Simplify away unnecessary qualifications.
+
+                var addImportService = document.GetLanguageService<IAddImportsService>();
+                ImmutableArray<SyntaxNode> containers;
+
+                if (refLocations.Count > 0)
+                {
+                    (document, containers) = await FixReferencesAsync(document, addImportService, _service, refLocations, _newNamespaceParts, cancellationToken)
+                        .ConfigureAwait(false);
+                }
+                else
+                {
+                    // If there's no reference to types declared in this document,
+                    // we will use root node as import container.
+                    containers = ImmutableArray.Create(await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false));
+                }
+
+                Debug.Assert(containers.Length > 0);
+
+                // Need to import all containing namespaces of old namespace and add them to the document (if it's not global namespace)
+                var namesToImport = CreateAllContainingNamespaces(_oldNamespaceParts);
+
+                var optionSet = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+                var placeSystemNamespaceFirst = optionSet.GetOption(GenerationOptions.PlaceSystemNamespaceFirst, document.Project.Language);
+                var documentWithAddedImports = await AddImportsInContainersAsync(
+                        document, 
+                        addImportService, 
+                        containers, 
+                        namesToImport, 
+                        placeSystemNamespaceFirst, 
+                        cancellationToken).ConfigureAwait(false);
+
+                var root = await documentWithAddedImports.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                root = _service.ChangeNamespaceDeclaration((TCompilationUnitSyntax)root, _oldNamespaceParts, _newNamespaceParts)
+                    .WithAdditionalAnnotations(Formatter.Annotation);
+
+                // Need to invoke formatter explicitly since we are doing the diff merge ourselves.
+                root = await Formatter.FormatAsync(root, Formatter.Annotation, documentWithAddedImports.Project.Solution.Workspace, optionSet, cancellationToken)
+                    .ConfigureAwait(false);
+
+                root = root.WithAdditionalAnnotations(Simplifier.Annotation);
+                var formattedDocument = documentWithAddedImports.WithSyntaxRoot(root);
+                return await Simplifier.ReduceAsync(formattedDocument, optionSet, cancellationToken).ConfigureAwait(false);
```

Comment:
is there a reason the reduce is called mutliple time rather than called once at the end?